### DISCUSSION TITLE
TeslaFi import problems

### DISCUSSION THREAD

**Discussion by Twiederh** [teslamate-org/teslamate#4392] [General]
Hi,
I have two issues when trying to import a TeslaFi backup into Teslamate:
the first is, that the vehicle-id in the export varies and looks the attached screenshot, the first occurrence looks like a binary value, the rest like values in ascending order.

When I replace the vehicle ID with a 1, I get the following error on the import:

{{:badmatch, {:error, #Ecto.Changeset< action: :insert, changes: %{ date: ~U[2024-11-12 14:15:44.000000Z], speed: 119, power: 43, latitude: Decimal.new("50.162142"), longitude: Decimal.new("8.91141"), usable_battery_level: 91, driver_temp_setting: Decimal.new("21.0"), fan_status: 1, inside_temp: Decimal.new("21.0"), outside_temp: Decimal.new("11.5"), passenger_temp_setting: Decimal.new("21.0"), odometer: 13.697001, ideal_battery_range_km: Decimal.new("475.69"), rated_battery_range_km: Decimal.new("475.69"), est_battery_range_km: Decimal.new("407.32") }, errors: [ battery_level: {"is invalid", [type: :integer, validation: :cast]} ], data: #TeslaMate.Log.Position<>, valid?: false, ... >}}, [ {TeslaMate.Vehicles.Vehicle, :"-handle_event/4-fun-3-", 4, [file: ~c"lib/teslamate/vehicles/vehicle.ex", line: 756]}, {Ecto.Adapters.SQL, :"-checkout_or_transaction/4-fun-0-", 3, [file: ~c"lib/ecto/adapters/sql.ex", line: 1400]}, {DBConnection, :run_transaction, 4, [file: ~c"lib/db_connection.ex", line: 1756]}, {TeslaMate.Vehicles.Vehicle, :handle_event, 4, [file: ~c"lib/teslamate/vehicles/vehicle.ex", line: 748]}, {:gen_statem, :loop_state_callback, 11, [file: ~c"gen_statem.erl", line: 1397]}, {:proc_lib, :init_p_do_apply, 3, [file: ~c"proc_lib.erl", line: 241]} ]}

I checked the lines mentioned in the error but they do not look suspicious.

This is the content of the log on the original import file:

2024-11-19 12:23:03.726 [error] GenStateMachine TeslaMate.Import terminating
** (Ecto.Query.CastError) deps/ecto/lib/ecto/repo/queryable.ex:487: value `"1 1 0 0 0 0 1"` in `where` cannot be cast to type :integer in query:
from c0 in TeslaMate.Log.Car,
  where: c0.vid == ^"1 1 0 0 0 0 1",
  select: c0
    (elixir 1.17.3) lib/enum.ex:2531: Enum."-reduce/3-lists^foldl/2-0-"/3
    (elixir 1.17.3) lib/enum.ex:1829: Enum."-map_reduce/3-lists^mapfoldl/2-0-"/3
2024-11-19 12:23:05.381 [info] GET /import
2024-11-19 12:23:05.386 [info] Sent 200 in 4ms
2024-11-19 12:23:07.652 [info] Importing 1 file(s) ...
2024-11-19 12:23:07.654 [info] Starting logger for 'Sams'
2024-11-19 12:23:07.656 [error] GenStateMachine TeslaMate.Import terminating
** (Ecto.Query.CastError) deps/ecto/lib/ecto/repo/queryable.ex:487: value `"1 1 0 0 0 0 1"` in `where` cannot be cast to type :integer in query:

Does anyone have a clue what is going wrong here?

Thanks in advance for your help.

Many regards,
Thomas.

**Comment by Twiederh** [comment]
<img width="109" alt="Bildschirmfoto 2024-11-19 um 14 47 04" src="https://github.com/user-attachments/assets/86dd89dc-1a13-45df-b4e3-5fb74f1232b3">
Here is the screenshot.

**Reply by Twiederh** [reply]
I was able to solve the problem with the help of someone at myteslamate: I copied the content of the usable_battery_level to the _battery_level column since teslafi expects an integer value for battery level. Then the import worked.

**Reply by tcbailey239** [reply]
I'm having issues with the import and noticed the same series of numbers in vehicle_id.  Your reply indicates you copied usable_battery_level to _battery_level and then it worked.  Assume this is just "battery_level" and not some other field I'm missing. Did you ever do anything with the vehicle_id field? 

I also notice the TeslaFi exports are named MMYYYY.csv with some using 1 character for the month (MYYYY.csv). The Teslamate instructions indicate the import filename has to be TeslaFiMMYYYY.csv although the import isn't finding the file unless I add a 0 to make it a full MMYYYY if that makes sense. Wondering if you had any issues there. Thank you.

**Comment by PharaohXev** [comment]
Hello,

I also have an issue while importing data from TeslaFi.
I saved all my data from January 2021 to January 2025. Then I imported them in TeslaMate and everything was OK (Green check) until December 2024. So the 2 last files, December 2024 and January 2025.

Displayed error is this one:
** (MatchError) no match of right hand side value: {:error, #Ecto.Changeset<action: :update, changes: %{eid: 52037}, errors: [vid: {"is invalid", [type: :integer, validation: :cast]}], data: #TeslaMate.Log.Car<>, valid?: false, ...>} 

I tested several times but I'm still unable to import those files.
I opened the last file which worked (November 2024) and compared to the December 2024 file and didn't see anything different.

When I try again to import them, the "import" button seems to be not working. No error, no "work in progress" icon, and when I go to the log files, I have the same error with different "eid: XXXXX".

Could you please help? 

Thank you!

**Reply by tcbailey239** [reply]
Hey @PharaohXev since you just did this, I'm running into some issues and hope you can provide some feedback. Did you have to make any modifications to the TeslaFi exports at all (either filename or vehicle_id)? I've tried several times today without luck.

**Reply by Twiederh** [reply]
You also need to change all rows in the vehicle_id column to 1 (or 2 if it is your second Tesla)

**Reply by PharaohXev** [reply]
@tcbailey239 I exported all the files at the same time, with the python script.
I also tried to re-export the files manually but none of the 2 operations succeeded to import the files :(
On all files, I changed nothing.
@Twiederh my vehicle_id column has a lot of different data:
![image](https://github.com/user-attachments/assets/4a223d44-e2ba-4f0c-977e-e8d91dad5a66)
But there are the same into the files which were uploaded without issues

**Comment by Xapier** [comment]
Is it possible that the import error has something to do with the data types? I received an error message that the battery_level column is faulty.

Upon closer inspection, I noticed that from around November 22, 2024, the data type changes from integer to decimal (e.g. 54 to 54.23). And for me, these are exactly the files from whose date I start getting errors when importing.

Unfortunately, I was not able to adjust the data en masse because Excel constantly ruins the CSV formatting and I have not found a CSV editor that can easily adjust several thousand rows.

Edit: Okay, its an issue already https://github.com/teslamate-org/teslamate/issues/4477

**Comment by JakobLichterfeld** [comment]
#5513 an #5519 improve Teslafi import a lot

**Reply by PharaohXev** [reply]
I just tested today to import again the 3 not imported files, and it still doesn't work. 
I have TeslaMate version 4.0.1

See the behavior:
https://github.com/user-attachments/assets/5d546c3a-86fd-4c91-beb2-639c9620e077

**Reply by JakobLichterfeld** [reply]
> I have TeslaMate version 4.0.1

That's expected behavior, you need to use edge version (which include the fixes since weeks) or wait some days till 4.1.0 will be released
